# Task
Analyze the "Loan-Approval-Prediction-Dataset" from Kaggle to build a classification model that predicts loan approval. Handle missing values, encode categorical features, and address class imbalance using techniques like SMOTE. Train and evaluate at least two classification models (e.g., Logistic Regression, Decision Tree) focusing on precision, recall, and F1-score, and compare their performance.

## Load the dataset

### Subtask:
Load the loan approval dataset into a pandas DataFrame.


**Reasoning**:
Import pandas, load the dataset, and display the head and info to understand the data.



**Reasoning**:
The previous attempt failed because the file was not found. I will try reading the file from a different location, assuming it might be in the '/data/' directory.



In [2]:
!pip install mlcroissant

  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 144.4/144.4 kB 2.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 565.1/565.1 kB 7.7 MB/s eta 0:00:00
  Created wheel for jsonpath-rw: filename=jsonpath_rw-1.4.0-py3-none-any.whl size=15127 sha256=d19c6dc579e9025a315f59133c376e858fbb349e8d7138458433438f8965f62e
  Stored in directory: /root/.cache/pip/wheels/e5/8d/50/ee73263c97069bd6040ff40633d444fefaac7beff73abe81a7
Successfully built jsonpath-rw


In [3]:
import mlcroissant as mlc
import pandas as pd

# Fetch the Croissant JSON-LD
croissant_dataset = mlc.Dataset('https://www.kaggle.com/datasets/mosaadhendam/loan-prediction-dataset/croissant/download')

# Check what record sets are in the dataset
record_sets = croissant_dataset.metadata.record_sets
print(record_sets)

# Fetch the records and put them in a DataFrame
record_set_df = pd.DataFrame(croissant_dataset.records(record_set=record_sets[0].uuid))
record_set_df.head()


  -  [Metadata(Loan Prediction Dataset)] Property "http://mlcommons.org/croissant/citeAs" is recommended, but does not exist.


[RecordSet(uuid="loan_prediction_dataset.csv")]


loan_prediction_dataset.csv/Age  loan_prediction_dataset.csv/Income  \
0                               56                               81788   
1                               69                              102879   
2                               46                               58827   
3                               32                              127188   
4                               60                               25655   

   loan_prediction_dataset.csv/Credit_Score  \
0                                       334   
1                                       781   
2                                       779   
3                                       364   
4                                       307   

   loan_prediction_dataset.csv/Loan_Amount  \
0                                    15022   
1                                    21013   
2                                    39687   
3                                    16886   
4                                    26256   

   loan_prediction_dataset.csv/Loan_Term  \
0                                     48   
1                                     24   
2                                     60   
3                                     24   
4                                     36   

  loan_prediction_dataset.csv/Employment_Status  \
0                                   b'Employed'   
1                              b'Self-Employed'   
2                              b'Self-Employed'   
3                                 b'Unemployed'   
4                                 b'Unemployed'   

   loan_prediction_dataset.csv/Loan_Approved  
0                                          0  
1                                          1  
2                                          0  
3                                          0  
4                                          0

In [4]:
display(record_set_df.isnull().sum())

loan_prediction_dataset.csv/Age                  0
loan_prediction_dataset.csv/Income               0
loan_prediction_dataset.csv/Credit_Score         0
loan_prediction_dataset.csv/Loan_Amount          0
loan_prediction_dataset.csv/Loan_Term            0
loan_prediction_dataset.csv/Employment_Status    0
loan_prediction_dataset.csv/Loan_Approved        0
dtype: int64

In [6]:
# Identify categorical columns
categorical_cols = record_set_df.select_dtypes(include='object').columns

# Apply one-hot encoding
record_set_df_encoded = pd.get_dummies(record_set_df, columns=categorical_cols, drop_first=True)

display(record_set_df_encoded.head())

loan_prediction_dataset.csv/Age  loan_prediction_dataset.csv/Income  \
0                               56                               81788   
1                               69                              102879   
2                               46                               58827   
3                               32                              127188   
4                               60                               25655   

   loan_prediction_dataset.csv/Credit_Score  \
0                                       334   
1                                       781   
2                                       779   
3                                       364   
4                                       307   

   loan_prediction_dataset.csv/Loan_Amount  \
0                                    15022   
1                                    21013   
2                                    39687   
3                                    16886   
4                                    26256   

   loan_prediction_dataset.csv/Loan_Term  \
0                                     48   
1                                     24   
2                                     60   
3                                     24   
4                                     36   

   loan_prediction_dataset.csv/Loan_Approved  \
0                                          0   
1                                          1   
2                                          0   
3                                          0   
4                                          0   

   loan_prediction_dataset.csv/Employment_Status_b'Self-Employed'  \
0                                              False                
1                                               True                
2                                               True                
3                                              False                
4                                              False                

   loan_prediction_dataset.csv/Employment_Status_b'Unemployed'  
0                                              False            
1                                              False            
2                                              False            
3                                               True            
4                                               True

# Task
Address class imbalance in the "Loan-Approval-Prediction-Dataset" dataset using techniques like SMOTE, split the data, train at least two classification models (e.g., Logistic Regression, Decision Tree), evaluate their performance using precision, recall, and F1-score, compare the models, and identify the best performing one.

## Address class imbalance

### Subtask:
Use techniques like SMOTE to handle the class imbalance in the target variable.


**Reasoning**:
Separate features and target, apply SMOTE to handle class imbalance.



In [7]:
from imblearn.over_sampling import SMOTE
from sklearn.model_selection import train_test_split

X = record_set_df_encoded.drop('loan_prediction_dataset.csv/Loan_Approved', axis=1)
y = record_set_df_encoded['loan_prediction_dataset.csv/Loan_Approved']

smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(X, y)

display(y_resampled.value_counts())

loan_prediction_dataset.csv/Loan_Approved
0    1658
1    1658
Name: count, dtype: int64

## Split data

### Subtask:
Split the data into training and testing sets.


**Reasoning**:
Split the resampled data into training and testing sets using train_test_split.



In [8]:
X_train, X_test, y_train, y_test = train_test_split(X_resampled, y_resampled, test_size=0.2, random_state=42)
print("Training set shapes:", X_train.shape, y_train.shape)
print("Testing set shapes:", X_test.shape, y_test.shape)

Training set shapes: (2652, 7) (2652,)
Testing set shapes: (664, 7) (664,)


## Train models

### Subtask:
Train at least two classification models (e.g., Logistic Regression, Decision Tree) on the training data.


**Reasoning**:
Import the necessary classes and train the two models.



In [9]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier

logistic_model = LogisticRegression(random_state=42)
decision_tree_model = DecisionTreeClassifier(random_state=42)

logistic_model.fit(X_train, y_train)
decision_tree_model.fit(X_train, y_train)

print("Logistic Regression model trained.")
print("Decision Tree model trained.")

Logistic Regression model trained.
Decision Tree model trained.


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


## Evaluate models

### Subtask:
Evaluate the performance of the trained models using precision, recall, and F1-score on the test data.


**Reasoning**:
Import necessary metrics, make predictions using the trained models, and calculate and print precision, recall, and F1-score for each model.



In [10]:
from sklearn.metrics import precision_score, recall_score, f1_score

logistic_pred = logistic_model.predict(X_test)
decision_tree_pred = decision_tree_model.predict(X_test)

print("Logistic Regression Model Evaluation:")
print("Precision:", precision_score(y_test, logistic_pred))
print("Recall:", recall_score(y_test, logistic_pred))
print("F1-score:", f1_score(y_test, logistic_pred))

print("\nDecision Tree Model Evaluation:")
print("Precision:", precision_score(y_test, decision_tree_pred))
print("Recall:", recall_score(y_test, decision_tree_pred))
print("F1-score:", f1_score(y_test, decision_tree_pred))

Logistic Regression Model Evaluation:
Precision: 0.7727272727272727
Recall: 0.9174603174603174
F1-score: 0.8388969521044993

Decision Tree Model Evaluation:
Precision: 1.0
Recall: 0.9968253968253968
F1-score: 0.9984101748807631


## Compare models

### Subtask:
Compare the performance metrics of the models and identify the best performing one.


## Summary:

### Data Analysis Key Findings

*   SMOTE was successfully applied to address the class imbalance in the target variable 'loan\_prediction\_dataset.csv/Loan\_Approved', resulting in a balanced dataset with 1658 instances for each class.
*   The resampled data was split into training (2652 samples) and testing (664 samples) sets.
*   Both Logistic Regression and Decision Tree models were trained on the balanced training data. A `ConvergenceWarning` was noted for the Logistic Regression model.
*   On the test set, the Logistic Regression model achieved a precision of approximately 0.773, recall of approximately 0.917, and an F1-score of approximately 0.839.
*   On the test set, the Decision Tree model achieved a precision of 1.0, recall of approximately 0.997, and an F1-score of approximately 0.998.

### Insights or Next Steps

*   The Decision Tree model significantly outperformed the Logistic Regression model across all evaluated metrics (precision, recall, and F1-score) on the test set.
*   Further investigation into the `ConvergenceWarning` for the Logistic Regression model, potentially by scaling the data or increasing `max_iter`, could be beneficial, although given the Decision Tree's performance, it may not be necessary if the goal is to select the best model from this set.


In [12]:
import joblib

# Save the Logistic Regression model
joblib.dump(logistic_model, 'logistic_regression_model.joblib')

# Save the Decision Tree model
joblib.dump(decision_tree_model, 'decision_tree_model.joblib')

print("Models saved successfully.")

Models saved successfully.
